### PR TITLE
Parquet Encryption

### PR DESCRIPTION
**Pull request by lnkuiper**
This PR implements encryption for Parquet files as an experimental feature. We mostly follow the [Parquet Modular Encryption specification](https://github.com/apache/parquet-format/blob/master/Encryption.md), save for some details. For now, this means that our Parquet encryption is not compatible with that of, e.g., PyArrow, until the missing details are implemented.

Named encryption keys of 128, 192, or 256 bits can be added to a session, which are stored in-memory:
```sql
PRAGMA add_parquet_key('key128', '0123456789112345');
PRAGMA add_parquet_key('key192', '012345678911234501234567');
PRAGMA add_parquet_key('key256', '01234567891123450123456789112345');
```

Files are then encrypted like so:
```sql
COPY tbl TO 'tbl.parquet' (ENCRYPTION_CONFIG {footer_key: 'key256'});
```

For now, we encrypt the footer and all columns using the `footer_key`. The Parquet specification allows encryption of individual columns with different keys, e.g.:
```sql
COPY tbl TO 'tbl.parquet' (ENCRYPTION_CONFIG {footer_key: 'key256',
                                              column_keys: {key256: ['col0', 'col1']}});
```
This will cause an error to be thrown (for now!):
```
Not implemented Error: Parquet encryption_config column_keys not yet implemented
```

The encrypted file can then be read like so:
```sql
COPY tbl FROM 'tbl.parquet' (ENCRYPTION_CONFIG {footer_key: 'key256'});
SELECT * FROM read_parquet('tbl.parquet', encryption_config={footer_key: 'key256'});
```

Encryption has some performance implications, of course. Without encryption, reading/writing the `lineitem` table from TPC-H at SF1, which is 6M rows and 15 columns, from/to a Parquet file takes 0.26s and 0.99s, respectively. With encryption, this takes 0.64s and 2.21s.

### CODE REVIEW

**Comment by carlopi** [comment]
I did check locally, and compiles with duckdb-wasm (was unsure about dependencies, and should set a proper CI job to do this). Yey!
Minor data point: DuckDB binary size difference (while bundling parquet, as per the default configuration) is 2%.

**Review by Tmonster — COMMENTED** [review]
Didn't look at the embedded libraries. Will also most likely look at this again after reading the [Parquet Modular Encryption](https://github.com/duckdb/duckdb/pull/9392/files#parquet-modular-encryption)

File: extension/parquet/include/parquet_crypto.hpp [hunk 1]
```diff
@@ -0,0 +1,87 @@
...
+
+class BufferedFileWriter;
+
+class ParquetKeys : public ObjectCacheEntry {
+public:
+	static ParquetKeys &Get(ClientContext &context);
+
+public:
+	void AddKey(const string &key_name, const string &key);
+	bool HasKey(const string &key_name) const;
```
[the remark this answers is not in the document]
Comment (Tmonster) [reply extension/parquet/include/parquet_crypto.hpp:30]: wait, I imagine this is just checking if the key is in cache or something right? using the `pragma add_encryption_key` or something

File: extension/parquet/parquet_crypto.cpp [hunk 2]
```diff
@@ -0,0 +1,370 @@
...
+		}
+
+		data_t read_tag[ParquetCrypto::TAG_BYTES];
+		transport_remaining -= trans.read(read_tag, ParquetCrypto::TAG_BYTES);
+		if (memcmp(computed_tag, read_tag, ParquetCrypto::TAG_BYTES) != 0) {
+			throw InvalidInputException("Computed AES tag differs from read AES tag, are you using the right key?");
+		}
+
+		if (transport_remaining != 0) {
+			throw InvalidInputException("Encoded ciphertext length differs from actual ciphertext length");
```
[the remark this answers is not in the document]
Comment (lnkuiper) [reply extension/parquet/parquet_crypto.cpp:214]: The length of the ciphertext (in bytes) is encoded right before the actual ciphertext. If, for some reason, the actual length differs from the encoded length, something is definitely wrong, and we have to throw an error.

File: extension/parquet/column_reader.cpp [hunk 3]
```diff
@@ -23,6 +23,8 @@
 #include "duckdb/common/types/chunk_collection.hpp"
 #endif
 
+#include <iostream>
```
[the remark this answers is not in the document]
Comment (lnkuiper) [reply extension/parquet/column_reader.cpp:26]: Nope, thanks!

**Comment by Youssef-Harby** [comment]
Perhaps this is not the right place to ask, but I'm unsure. Can encryption be applied on a column-by-column basis, such that specific information is accessible only with a specific encryption key?
I mean not only the whole file 

Maybe attribute based authorization or Attribute-Based Access Control (ABAC)

**Comment by Tmonster** [comment]
Hi @Youssef-Harby,

Unfortunately we do not support encryption on a column by column basis yet. 
Feel free to take a look at the docs for any other questions! https://duckdb.org/docs/data/parquet/encryption

**Comment by alexandroid11** [comment]
Hello,

Doing very simple test for .10 version
(Downloaded latest version with winget install DuckDB.cli )

select * from md;
┌────────┐
│   Id   │
│ int32  │
├────────┤
│ 0 rows │
└────────┘
 insert into md (Id) values (1);
 insert into md (Id) values (2);
 select * from md;
┌───────┐
│  Id   │
│ int32 │
├───────┤
│     1 │
│     2 │
└───────┘
 PRAGMA add_parquet_key('key256', '01234567891123450123456789112345');
┌─────────┐
│ Success │
│ boolean │
├─────────┤
│ 0 rows  │
└─────────┘
 COPY md TO 'tbl.parquet' (ENCRYPTION_CONFIG {footer_key: 'key256'});

Error: Invalid Error: Unable to generate random data

Not sure what I'm doing wrong.

**Comment by szarnyasg** [comment]
@alexandroid11 this works fine for me on macOS:
```sql
create table md (id int32);
select * from md;
insert into md (Id) values (1);
insert into md (Id) values (2);
select * from md;
PRAGMA add_parquet_key('key256', '01234567891123450123456789112345');
COPY md TO 'tbl.parquet' (ENCRYPTION_CONFIG {footer_key: 'key256'});
```
```sql
PRAGMA add_parquet_key('key256', '01234567891123450123456789112345');
FROM read_parquet('tbl.parquet', encryption_config = {footer_key: 'key256'});
```
```
┌───────┐
│  id   │
│ int32 │
├───────┤
│     1 │
│     2 │
└───────┘
```

**Comment by szarnyasg** [comment]
If you can reproduce the issue on a fresh set up, please feel free to open a new issue!

**Pull request merged**